File: scripts/slim_mcdata.py
#!/usr/bin/env python3
"""Slim minecraft-data/data.js untuk build browser (CIVITAS OS v1.5).
Menyimpan HANYA versi pc yang dipakai klien web (1.21.1 utama, 1.20.4 cadangan),
membackup original ke data.js.orig. Idempoten: selalu bekerja dari .orig.
"""
import os
import re

D = "/home/z/my-project/node_modules/minecraft-data/data.js"
ORIG = D + ".orig"
KEEP = ["1.21.1", "1.20.4"]

if not os.path.exists(ORIG):
    with open(D, encoding="utf8") as f:
        src = f.read()
    with open(ORIG, "w", encoding="utf8") as f:
        f.write(src)
    print("original dibackup -> data.js.orig")
else:
    with open(ORIG, encoding="utf8") as f:
        src = f.read()

# potong blok per versi pc: '    <versi>: {' ... sampai '    },' atau '    }'
lines = src.split("\n")
out = []
in_pc = False
keep_this = False
depth_buf: list[str] = []
kept_count = 0
dropped = 0
for ln in lines:
    m_pc = re.match(r"^  'pc': \{$", ln)
    if m_pc:
        in_pc = True
        out.append(ln)
        continue
    if in_pc:
        m_ver = re.match(r"^    '([^']+)': \{$", ln)
        if m_ver:
            keep_this = m_ver.group(1) in KEEP
            if keep_this:
                kept_count += 1
            else:
                dropped += 1
            out.append(ln)
            continue
        if re.match(r"^  \},$", ln) or re.match(r"^  \}$", ln):
            in_pc = False
            keep_this = False
            out.append(ln)
            continue
        # di dalam blok versi: baris isi hanya dipertahankan bila versi disimpan
        if keep_this:
            out.append(ln)
        elif ln.strip() == "":
            out.append(ln)
        # selain itu dibuang (isi versi yang di-drop)
        continue
    out.append(ln)

# bersihkan koma ganda yang mungkin muncul (blok terakhir)
result = "\n".join(out)
result = re.sub(r",(\s*\},\s*\n  \},)", r"\1", result)

with open(D, "w", encoding="utf8") as f:
    f.write(result)
print(f"versi dipertahankan: {kept_count}, dibuang: {dropped}")

File: scripts/slim_mcdata.js
#!/usr/bin/env node
// Regenerasi data.js minecraft-data: hanya versi pc tertentu (hemat memori build).
// Sumber: data.js.orig (backup asli). Idempoten.
const fs = require("fs");
const path = require("path");

const D = "/home/z/my-project/node_modules/minecraft-data/data.js";
const ORIG = D + ".orig";
const KEEP = ["1.21.1", "1.20.4"];

if (!fs.existsSync(ORIG)) {
  fs.copyFileSync(D, ORIG);
  console.log("backup dibuat: data.js.orig");
}

// bersihkan cache agar load .orig (bukan slim sebelumnya)
const orig = require(ORIG);
const pc = orig.pc || {};
let body = "";
for (const ver of KEEP) {
  if (!pc[ver]) { console.error("versi tidak ada di data asli:", ver); process.exit(1); }
  const props = [];
  const src = Object.getOwnPropertyNames(pc[ver]);
  for (const key of src) {
    const desc = Object.getOwnPropertyDescriptor(pc[ver], key);
    if (desc.get) {
      // getter: temukan jalur require asli dari toString getter
      const fn = String(desc.get);
      const m = fn.match(/require\("([^"]+)"\)/);
      if (!m) { console.error("getter tanpa require:", ver, key); process.exit(1); }
      props.push(`      get ${key} () { return require(${JSON.stringify(m[1])}) }`);
    } else if (key === "proto") {
      props.push(`      proto: __dirname + ${JSON.stringify(desc.value.slice(desc.value.indexOf("/minecraft-data")))}`);
    } else {
      props.push(`      ${key}: ${JSON.stringify(desc.value)}`);
    }
  }
  body += `    ${JSON.stringify(ver)}: {\n${props.join(",\n")}\n    },\n`;
}
const out = "//patched + slimmed by CIVITAS OS (hanya versi yang dipakai klien web)\nmodule.exports =\n{\n  'pc': {\n" + body + "  },\n}\n";
fs.writeFileSync(D, out);
// verifikasi
delete require.cache[require.resolve(D)];
const chk = require(D);
const ok = KEEP.every((v) => chk.pc[v] && typeof chk.pc[v].protocol === "object" && typeof chk.pc[v].blocks === "object");
console.log("versions:", Object.keys(chk.pc).join(","), "| 1.21.1 ok:", ok, "| size:", (fs.statSync(D).size / 1024).toFixed(0) + "KB");
